docs/http-status-codes.adoc

```diff
@@ -0,0 +1,166 @@
+* the request to create the instance has been received and we're working on it ("202 Accepted"), but the resource won't show up until the Instance is running
+
+We opt for the first approach in most cases because then you can then fetch the Instance again and see its state, etc.
+
+In cases where the request does not create a new resource (e.g., "detach a disk"), we use "202 Accepted".
+
+=== Client errors vs. server errors
+
+When it comes to errors, **the most important distinction is between 400-level and 500-level status codes**.  https://httpwg.org/specs/rfc7231.html#status.codes[RFC 7231] summarizes the distinction:
+
+* 400-level codes (often called "client errors") mean "the request contains bad syntax or cannot be fulfilled".
+* 500-level codes (often called "server errors") mean "the server failed to fulfill an apparently valid request".
+
+Critically, a 500-level response means the problem is _completely outside the client's control_.  Common reasons include that an internal dependency is offline or overloaded or the server just hit a bug.  There's nothing the client can do except maybe retry, and even that isn't always appropriate.
+
+400-level codes don't necessarily represent a problem, mistake, or bug.  A client might attempt a conditional GET and get back a 400-level error saying the preconditions weren't true.  That might be totally expected on the client side under normal conditions.
+
+=== Client errors
+
+We use "400 Bad Request" for most types of invalid input.  This is pretty arbitrary.  Other popular choices include "409 Conflict" and "422 Unprocessible entity".  The spec for "409" really only seems to apply when the requested change conflicts with the underlying state, which is some kinds of invalid input (e.g., booting an Instance that's currently running) but not all.  422 comes from WebDAV and is not even mentioned in RFC 7231.
+
+=== Server errors
+
+Some clients erroneously treat all 500-level errors the same and retry them.  But it's useful to distinguish them:
+
+"500 Internal Server Error" basically means that the server hit a _bug_.  Retrying is not likely to be useful.  If an operator sees this, they should probably call support.  We might automatically open a support case when we see this.
+
+"503 Service Unavailable" means the server is _currently_ unable to handle the request -- usually this means something is overloaded or a dependency is not working.  Retrying _is_ likely a good idea.  On the operator side: the system should be providing the operator with information about known problems that might cause this, like sleds that are offline or disks that need to be replaced.  If the operator has resolved all the issues being reported them and they still see 503s, that's a bug!
+
+=== Other status codes
+
+There are also https://httpwg.org/specs/rfc7231.html#rfc.section.6.2[100-level status codes], whose use is generally more standardized and well-understood.
```

Comment:
would you imagine us using these?

Reply:
Yup.  The two I know of are:

- 100-continue, which is used when the client uses the "expect: 100-continue" header on a PUT.  This is a way of saying "please validate the meat of this request before I start uploading [a gigantic file] [so I don't only find out after uploading the whole thing that the request is going to fail]".  If we ever support uploads of anything (e.g., update images via offline media), we may want this.
- 100 "switching protocols": used for websockets, which I think we're planning to use for serial console and VNC.

While these are kind of obscure, I think we may well use them.  And if we do, I expect it will be in the totally normal way that they're generally used, so I didn't want to re-explain it here.